File: manual/java-src/com/atta/mcpanel/overlay/OverlayService.java
package com.atta.mcpanel.overlay;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Build;
import android.os.IBinder;

/** سرویس پیش‌زمینهٔ پنل شناور روی بازی */
public class OverlayService extends Service {

    public static final String CHANNEL_ID = "atta_overlay";
    private static final int NOTIF_ID = 1001;

    private static volatile boolean running = false;

    private PanelHost host;
    private boolean cfgRegistered = false;

    /** چرخش/تغییر نوار سیستم → جای‌گذاری مجدد پنل داخل کادر */
    private final BroadcastReceiver configReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (host != null) host.onConfigChanged();
        }
    };

    public static boolean isRunning() { return running; }

    public static void start(Context context) {
        if (running) return;
        running = true;
        try {
            context.startForegroundService(new Intent(context, OverlayService.class));
        } catch (Exception e) {
            running = false;
        }
    }

    public static void stop(Context context) {
        running = false;
        context.stopService(new Intent(context, OverlayService.class));
    }

    private void registerConfigReceiver() {
        if (cfgRegistered) return;
        try {
            IntentFilter f = new IntentFilter(Intent.ACTION_CONFIGURATION_CHANGED);
            if (Build.VERSION.SDK_INT >= 33) {
                // Context.RECEIVER_NOT_EXPORTED = 0x4 (ثابت در API 33 اضافه شده)
                registerReceiver(configReceiver, f, 0x4);
            } else {
                registerReceiver(configReceiver, f);
            }
            cfgRegistered = true;
        } catch (Exception ignored) {
            cfgRegistered = false;
        }
    }

    private void unregisterConfigReceiver() {
        if (!cfgRegistered) return;
        try {
            unregisterReceiver(configReceiver);
        } catch (Exception ignored) {}
        cfgRegistered = false;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        running = true;
        NotificationManager nm = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
        NotificationChannel ch = new NotificationChannel(
                CHANNEL_ID, "پنل شناور روی بازی", NotificationManager.IMPORTANCE_LOW);
        ch.setShowBadge(false);
        nm.createNotificationChannel(ch);
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        startInForeground();
        registerConfigReceiver();
        if (host == null) {
            host = new PanelHost(this);
            host.attach();
        } else {
            host.onConfigChanged();
        }
        return START_STICKY;
    }

    private void startInForeground() {
        Intent open = new Intent(this, com.atta.mcpanel.MainActivity.class);
        PendingIntent pi = PendingIntent.getActivity(
                this, 1, open,
                PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);

        Intent stopIntent = new Intent(this, OverlayReceiver.class);
        PendingIntent stopPi = PendingIntent.getBroadcast(
                this, 2, stopIntent,
                PendingIntent.FLAG_UPDATE_CURRENT | PendingIntent.FLAG_IMMUTABLE);

        int iconId = getResources().getIdentifier("ic_stat", "drawable", getPackageName());
        Notification n = new Notification.Builder(this, CHANNEL_ID)
                .setSmallIcon(iconId)
                .setContentTitle("پنل اپراتور Atta فعال است")
                .setContentText("نوار سبز لبهٔ چپ را لمس کن")
                .setContentIntent(pi)
                .addAction(0, "توقف پنل", stopPi)
                .setOngoing(true)
                .setCategory(Notification.CATEGORY_SERVICE)
                .build();
        startForeground(NOTIF_ID, n);
    }

    @Override
    public void onDestroy() {
        unregisterConfigReceiver();
        if (host != null) {
            host.detach();
            host = null;
        }
        running = false;
        super.onDestroy();
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: app/src/main/java/com/atta/mcpanel/overlay/OverlayService.kt
package com.atta.mcpanel.overlay

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.content.pm.ServiceInfo
import android.os.Build
import android.os.IBinder
import androidx.core.app.NotificationCompat
import com.atta.mcpanel.MainActivity
import com.atta.mcpanel.R

/**
 * سرویس پیش‌زمینهٔ پنل شناور روی بازی.
 * پنجرهٔ overlay را در [PanelHost] نگه می‌دارد و یک اعلان همیشگی نمایش می‌دهد.
 */
class OverlayService : Service() {

    private var host: PanelHost? = null

    override fun onCreate() {
        super.onCreate()
        running = true
        createChannel()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        startInForeground()
        if (host == null) {
            host = PanelHost(this)
            host?.attach()
        }
        return START_STICKY
    }

    private fun startInForeground() {
        val pi = PendingIntent.getActivity(
            this, 1,
            Intent(this, MainActivity::class.java),
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
        val stopPi = PendingIntent.getBroadcast(
            this, 2,
            Intent(OverlayReceiver.ACTION_STOP).setPackage(packageName),
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
        val n: Notification = NotificationCompat.Builder(this, CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_stat)
            .setContentTitle(getString(R.string.notif_title))
            .setContentText(getString(R.string.notif_text))
            .setContentIntent(pi)
            .addAction(0, getString(R.string.notif_stop), stopPi)
            .setOngoing(true)
            .setCategory(NotificationCompat.CATEGORY_SERVICE)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .build()
        if (Build.VERSION.SDK_INT >= 34) {
            startForeground(NOTIF_ID, n, ServiceInfo.FOREGROUND_SERVICE_TYPE_SPECIAL_USE)
        } else {
            startForeground(NOTIF_ID, n)
        }
    }

    override fun onDestroy() {
        host?.detach()
        host = null
        running = false
        super.onDestroy()
    }

    override fun onBind(intent: Intent?): IBinder? = null

    private fun createChannel() {
        val mgr = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val ch = NotificationChannel(
            CHANNEL_ID,
            getString(R.string.channel_name),
            NotificationManager.IMPORTANCE_LOW,
        ).apply {
            description = getString(R.string.channel_desc)
            setShowBadge(false)
        }
        mgr.createNotificationChannel(ch)
    }

    companion object {
        private const val CHANNEL_ID = "atta_overlay"
        private const val NOTIF_ID = 1001

        @Volatile
        var running: Boolean = false
            private set

        fun isRunning(): Boolean = running

        fun start(context: Context) {
            if (running) return
            running = true // خوش‌بینانه؛ رابط بلافاصله به‌روز شود
            try {
                context.startForegroundService(Intent(context, OverlayService::class.java))
            } catch (e: Exception) {
                running = false
            }
        }

        fun stop(context: Context) {
            running = false
            context.stopService(Intent(context, OverlayService::class.java))
        }
    }
}
